File: Assets/src/model/tasks/MoveNextToTargetTask.cs
using System.Collections.Generic;
using System.Linq;
using UnityEngine;

[System.Serializable]
public class MoveNextToTargetTask : FollowPathTask {
  public MoveNextToTargetTask(Actor actor, Vector2Int target) : base(actor, target, FindBestAdjacentPath(actor.pos, target)) { }

  public static List<Vector2Int> FindBestAdjacentPath(Vector2Int pos, Vector2Int target) {
    if (pos == target) {
      return new List<Vector2Int>();
    }
    var path = GameModel.main.currentFloor.FindPath(pos, target, true);
    if (path.Any()) {
      path.RemoveAt(path.Count - 1);
    }
    return path;
  }
}

File: web/src/model/tasks/MoveNextToTargetTask.ts
import { FollowPathTask } from './FollowPathTask';
import { Vector2Int } from '../../core/Vector2Int';
import { GameModelRef } from '../GameModelRef';
import type { Actor } from '../Actor';
import type { Body } from '../Body';

export class MoveNextToTargetTask extends FollowPathTask {
  constructor(actor: Actor, target: Vector2Int) {
    super(actor, target, MoveNextToTargetTask.findBestAdjacentPath(actor, target));
  }

  static findBestAdjacentPath(body: Body, target: Vector2Int): Vector2Int[] {
    if (Vector2Int.equals(body.pos, target)) return [];
    const floor = GameModelRef.main.currentFloor;
    const path = floor.findPath(body.pos, target, true, body);
    if (path.length > 0) {
      path.pop(); // Remove the target itself — we want to be *next to* it
    }
    return path;
  }
}
